### ISSUE TITLE
Implement virtual scrolling 

### ISSUE THREAD

**Issue by Biont** [owncloud/music#598]
The current UI does not scale very well with large collections. This is because all artists and all albums are rendered to the DOM, no matter if the user is ever going to see them all. Plus, there are a number of event handlers on every one of them. This creates a lot of overhead the larger your music collection grows.

There are solutions to handle large lists effectively, like ionic collection repeat and [angular-vs-repeat](https://github.com/kamilkp/angular-vs-repeat).

Using these, the collection handler will ensure that only the visible segment of the list is actually being rendered, adding and removing elements as needed. This would make rendering performance in the music app a non-issue. (and you also get lazy-loaded cover art for free, since it won't be downloaded until you scroll it into view)

I have a working prototype using angular-vs-repeat in my fork and would love to PR it when it matures. (I hope to be able to push it soon, it's still a but ugly due to missing styling because of markup changes)

The only downside is that we would need to change the layout. For virtual scrolling to work, you need to know the exact height of the list (and all individual items). This is almost impossible with the current layout, since there can be 1 or dozens of albums for a single artist that reflow depending on the browser window width.

We need a fixed-height list. If elements can change size, it needs to be done with JS, so that we can have the collection manager calculate new offsets.

I could imagine having something similar to the collapsible track list next to an album: A fixed-size Artist with a fixed number of albums, and then a button that displays all of them if there are more.

What do you think?

**Comment by paulijar** [comment]
The idea of virtual scrolling has been recorded for the first time in #78 some four years ago but so far it has not happened. I did some prestudy on the virtual scrolling libraries some half a year ago but came to the same conclusion as you: the current UI design of the Music app does not fit the virtual scrolling. As I didn't want to redesign the app, I then dropped the idea.

I kind of like the current UI design and how it adapts to different screen/window sizes. Also, in my opinion, the performance is nowadays quite sufficient on music collections with up to 10k tracks. The real problems start on collections with tens of thousands of tracks. But my gut feeling is that for the vast majority of the potential user base the 10k tracks is enough. So the question is, is it sensible to sacrifice even a little on the usability of the average user just to better serve the few heavy-duty users.

Everything I have said so far is about the "Albums" view, but the "All tracks" view (and other playlist views) is a different story. It has a much simpler design, and using virtual scrolling on it has been on my mental to-do list for some time. We would only need to sacrifice the ability to wrap a single track to multiple lines which may happen on narrow (mobile device) screens. But I guess we can live without it. Using virtual scrolling on the All tracks view should make a big difference on the time it takes to swap between the different views even if the Albums view was not altered at all.

Another idea, which I've been pondering, is to make kind of virtual scrolling on the Albums view manually without relying on the libraries. It could work so that all the artists and albums would be added to the UI immediately, but the tracks would be injected to the albums only once the album is scrolled to the screen. In a typical library, the bulk of all the DOM elements comes from the tracks and rendering only the albums should take only a small fraction of the time it takes to render the whole library.

**Comment by Biont** [comment]
You make good points.

> So the question is, is it sensible to sacrifice even a little on the usability of the average user just to better serve the few heavy-duty users.

This got me thinking. Yes, as a general rule, you should not take away functionality/usability for under-the-hood improvements that solve problems which most users might never encounter.
If, however, you can change the current state into something that's _just as good_ to get those improvements, then I'd say it's justified. I think we can agree that virtual scrolling is an objectively better approach to rendering large lists, it's just that you're not willing to sacrifice usability for it. I realize now that you are right: My initial idea of creating fold-out albums is inferior and introduces extra clicks with no tangible benefit for the average user. Scratch that.

I think we can do something that's _just as good_ : Replace the completely fluid "album list" (which simply breaks lines when it runs out of space) with a small grid. Then you have a fixed number of columns for mobile/tablet/desktop/ultrawide, giving us much more control over how the list will look in all browser widths. 

Now, the whole layout is still very dynamic, but in a predictable manner that can be precomputed:

```js

// This is completeley off the top of my head, so it might not work, but should give you the right idea

var artist; // we got this from somewhere in the overviewcontroller
var albumHeight = 300; // this is fixed for the sake of simplicity. could be breakpoint-dependent as well

artist.dimensions = {};

// How many columns per row on each possible layout
var cols = {
 phone: 2,
 tablet: 4,
 desktop: 6,
 ultrawide: 10
}

for (var breakpoint in cols) {
  var numAlbums = artist.albums.length;
  var rows = albumHeight * ( Math.floor( numAlbums / cols[breakpoint] ) );
  if( ! numAlbums % cols[breakpoint] ){ // Add a row for remaining albums
     rows++;
  }
  artist.dimensions[breakpoint]  = rows * albumHeight;
}
```

Virtual scrolling does not need to be "fixed size", you just need an efficient way to get each item's total height. This now possible since we just have to dynamically tell the collection manager which dimension to use for which window size.  This can probably be done in a very simple custom angular directive.

With this solution, the needed UI changes would be absolutely minimal and we would still end up with a very very efficient list. Sounds good?

Virtual scrolling for the Tracks view should be trivial. If we and up with a PR here, I will make sure to include it.

**Comment by paulijar** [comment]
I agree, if the virtual scrolling can be applied without significant functional UI changes, then it is definitely a good improvement. And the grid-like layout might be a viable option to implement it. We just need to make sure that also the expanding/collapsing of the album track lists works nicely, taking into account that the expanded list may have 7 tracks or 100+ tracks or something in between.

I have a feeling that altogether this is a rather large job. But if you can pull it off, a PR is greatly appreciated.

**Comment by paulijar** [comment]
Closing, as a kind of custom-made virtual-scrolling has been released in v0.12.1. There, the virtual scrolling is applied to Albums and Folders views, and in v0.13.0, it will be extended to the All tracks view, too.
